### ISSUE TITLE
shellcheck: user error (Text.Regex.Posix.String died: (ReturnCode 17,"illegal byte sequence"))

### ISSUE THREAD

**Issue by apennebaker-ni** [koalaman/shellcheck#339]
When I try to shellcheck my wget-memegenerator script, ShellCheck fails on a regex error.

**Trace**

```
$ shellcheck lib/wget-memegenerator 
shellcheck: user error (Text.Regex.Posix.String died: (ReturnCode 17,"illegal byte sequence"))
```

Update:

I narrowed down the problem to the ASCII `»` symbol in:

```
PAGES=$(wget -q -O- "http://memegenerator.net/$TAG/images/new/alltime/page/1" | grep » -B 6 | grep page | head -n 1 | tr "[<>]" " " | awk '{ print $3 }')
```

That's where shellcheck's regex fails.

**Source**

https://github.com/mcandre/politerips/blob/shellcheck-339/lib/wget-memegenerator#L13

**System**

```
$ specs shellcheck cabal ghc brew os
Specs:

specs 0.14
https://github.com/mcandre/specs#readme

shellcheck --version
ShellCheck - shell script analysis tool
version: 0.3.6
license: GNU Affero General Public License, version 3
website: http://www.shellcheck.net

cabal --version
cabal-install version 1.22.0.1
using version 1.22.1.0 of the Cabal library 

ghc --version
The Glorious Glasgow Haskell Compilation System, version 7.8.4

brew --version
0.9.5

system_profiler SPSoftwareDataType | grep 'System Version'
      System Version: OS X 10.10.2 (14C1514)
```

**Mitigation**

For now, I'm working around this shellcheck regex parsing issue by using a hexcode to represent the offending character.

https://github.com/mcandre/politerips/commit/517bf9299f38ebe2d6a47a7d8705654b2a7ff0eb

**Comment by dualbus** [comment]
Can you provide us with a copy of the script that shows the problem?

-- 
Eduardo Bustamante
https://dualbus.me/

**Comment by koalaman** [comment]
This looks like #284, which I was unable to reproduce. Could you:

1. Post output of `locale` 
2. Post hex dump of the shortest script that reproduces the issue
3. Try again with a locale from `locale -a` whose encoding matches that of the script (if that's not the case already)?

**Comment by apennebaker-ni** [comment]
```
$ locale
LANG="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_CTYPE="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_ALL=

$ file -I lib/wget-memegenerator 
lib/wget-memegenerator: text/x-shellscript; charset=utf-8
```

As I said, the character that triggers the error is the » in `PAGES=$(wget -q -O- "http://memegenerator.net/$TAG/images/new/alltime/page/1" | grep » -B 6 | grep page | head -n 1 | tr "[<>]" " " | awk '{ print $3 }')`

**Comment by koalaman** [comment]
That takes care of 1, and from `file` output I'll just assume that the answer to 2 is `C2 BB`

I'm still unable to reproduce it with my system's en_US utf-8 locale (`en_US.utf8` as listed by `locale -a`). 

    vidar@vidarholen-vm ~/tmp $ cat script
    PAGES=$(wget -q -O- "http://memegenerator.net/$TAG/images/new/alltime/page/1" | grep » -B 6 | grep page | head -n 1 | tr "[<>]" " " | awk '{ print $3 }')

    vidar@vidarholen-vm ~/tmp $ file -i script
    script: text/plain; charset=utf-8

    vidar@vidarholen-vm ~/tmp $ locale
    LANG=en_US.utf8
    LANGUAGE=
    LC_CTYPE="en_US.utf8"
    LC_NUMERIC="en_US.utf8"
    LC_TIME="en_US.utf8"
    LC_COLLATE="en_US.utf8"
    LC_MONETARY="en_US.utf8"
    LC_MESSAGES="en_US.utf8"
    LC_PAPER="en_US.utf8"
    LC_NAME="en_US.utf8"
    LC_ADDRESS="en_US.utf8"
    LC_TELEPHONE="en_US.utf8"
    LC_MEASUREMENT="en_US.utf8"
    LC_IDENTIFICATION="en_US.utf8"
    LC_ALL=en_US.utf8

    $ shellcheck script
    In script line 1:
    PAGES=$(wget -q -O- "http://memegenerator.net/$TAG/images/new/alltime/page/1" | grep » -B 6 | grep page | head -n 1 | tr "[<>]" " " | awk '{ print $3 }')
    ^-- SC2148: Tips depend on target shell and yours is unknown. Add a shebang.
    ^-- SC2034: PAGES appears unused. Verify it or export it.
                                                                                                                             ^-- SC2021: Don't use [] around ranges in tr, it replaces literal square brackets.

This is with regex-compat version 0.95.1

**Comment by apennebaker-ni** [comment]
Hmm, try the larger script. We can pare it down to a minimal case.

https://github.com/mcandre/politerips/blob/shellcheck-339/lib/wget-memegenerator

**Comment by koalaman** [comment]
Works fine when I `wget https://raw.githubusercontent.com/mcandre/politerips/shellcheck-339/lib/wget-memegenerator`

    $ file -I wget-memegenerator 
    file: invalid option -- 'I'

    $ file -i wget-memegenerator 
    wget-memegenerator: text/plain; charset=utf-8

    $ shellcheck wget-memegenerator 
    In wget-memegenerator line 13:
    PAGES=$(wget -q -O- "http://memegenerator.net/$TAG/images/new/alltime/page/1" | grep » -B 6 |     grep page | head -n 1 | tr "[<>]" " " | awk '{ print $3 }')
                                                                                                                             ^-- SC2021: Don't use [] around ranges in tr, it replaces literal square brackets.


Here's parts of `cabal configure -v`:

    Configuring ShellCheck-0.3.6...
    Dependency QuickCheck ==2.7.6: using QuickCheck-2.7.6
    Dependency ShellCheck -any: using ShellCheck-0.3.6
    Dependency base ==4.5.0.0: using base-4.5.0.0
    Dependency containers ==0.4.2.1: using containers-0.4.2.1
    Dependency directory ==1.2.1.0: using directory-1.2.1.0
    Dependency json ==0.5: using json-0.5  
    Dependency mtl ==2.1.1: using mtl-2.1.1
    Dependency parsec ==3.1.2: using parsec-3.1.2
    Dependency regex-compat ==0.95.1: using regex-compat-0.95.1
    Dependency transformers ==0.3.0.0: using transformers-0.3.0.0

Which OS/distro and release are you using?

**Comment by apennebaker-ni** [comment]
Well, the funny thing is, I'm using ShellCheck in a non-Haskell project, so I haven't even written a cabal file yet. Could you hand me a minimal cabal file for installing ShellCheck in a sandbox for politerips?

```
$ cabal configure -v
/usr/local/bin/ghc --numeric-version
looking for tool ghc-pkg near compiler in /usr/local/bin
found ghc-pkg in /usr/local/bin/ghc-pkg
/usr/local/bin/ghc-pkg --version
/usr/local/bin/ghc --supported-languages
/usr/local/bin/ghc --info
Reading available packages...
cabal: No cabal file found.
Please create a package description file <pkgname>.cabal
```

System:

```
specs cabal ghc os
Specs:

specs 0.14
https://github.com/mcandre/specs#readme

cabal --version
cabal-install version 1.22.0.1
using version 1.22.1.0 of the Cabal library 

ghc --version
The Glorious Glasgow Haskell Compilation System, version 7.8.4

system_profiler SPSoftwareDataType | grep 'System Version'
      System Version: OS X 10.10.2 (14C1514)
```

**Comment by koalaman** [comment]
I was configuring the ShellCheck source code itself. dualbus found [a stackoverflow post](http://stackoverflow.com/questions/5047626/matching-specific-unicode-char-in-haskell-regexp) suggesting that it's specific to OS X, which would certainly explain why I've never been able to reproduce it.

This may just be a bug in the aging `regex-compat`. I'll look into newer regex libraries.

**Comment by apennebaker-ni** [comment]
Thanks for investigating! I love ShellCheck, hopefully a newer lib could fix dis.

**Comment by koalaman** [comment]
I was able to confirm the problems with regex-compat independently of ShellCheck on OS X. Apparently that's not the only problem with this package. regex-tdfa worked, so I upgraded to use it in 9f1f00c. 

Give it another go and see if it helped.

**Comment by apennebaker-ni** [comment]
Could you publish this as a bugfix version bump to make it easier to install?

**Comment by koalaman** [comment]
@apennebaker-ni Done. 0.3.7 includes this change plus fixes to compile on GHC 7.10.
